cuda/matrix/dense_kernels.cu

```diff
@@ -334,8 +334,10 @@ void compute_norm2(std::shared_ptr<const CudaExecutor> exec,
 {
     if (cublas::is_supported<ValueType>::value) {
         auto handle = cublas::init();
-        cublas::norm2(handle, result->get_size()[1], result->get_values(),
-                      result->get_stride(), result->get_values());
+        for (size_type col = 0; col < x->get_size()[1]; ++col) {
+            cublas::norm2(handle, x->get_size()[0], x->get_const_values() + col,
+                          x->get_stride(), result->get_values() + col);
```

Comment:
If I'm reading this correctly, `result->get_values() + col` is a `ValueType *`. In case `ValueType = std::complex<T>` (for `T` either `float` or `double`), which cuBLAS function does `cublas::norm2` map to? According to the documentation we have these options:

```c++
cublasStatus_t  cublasSnrm2(cublasHandle_t handle, int n,
                            const float           *x, int incx, float  *result);
cublasStatus_t  cublasDnrm2(cublasHandle_t handle, int n,
                            const double          *x, int incx, double *result);
cublasStatus_t cublasScnrm2(cublasHandle_t handle, int n,
                            const cuComplex       *x, int incx, float  *result);
cublasStatus_t cublasDznrm2(cublasHandle_t handle, int n,
                            const cuDoubleComplex *x, int incx, double *result);
```

Neither of them has a complex return value.

I think that the result should actually be of type `matrix::Dense<remove_complex<ValueType>>`, and we should ensure that the `result` is a real version of `ValueType` anywhere `Dense<ValueType>::compute_norm2` is used.

Reply:
With what I did, this should normally fail because I specified the overload of norm2 only with float/double ValueType for result, except if I did something wrong. So currently if result is complex it should throw the `not_implemented` exception.

See https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/base/cublas_bindings.hpp

Reply:
But I guess it is more correct to put result as `matrix::Dense<remove_complex<ValueType>>` indeed.

Reply:
> With what I did, this should normally fail because I specified the overload of norm2 only with float/double ValueType for result, except if I did something wrong. So currently if result is complex it should throw the `not_implemented` exception.

Well this is kind of a problem, because instead of fixing the problem with the wrong stopping criteria (which had a simple user-side workaround of just doubling the requested residual norm), it completely breaks it. If I now try to run the inverse iteration example with the CUDA executor I get this:

```
This is Ginkgo 0.0.0 (develop)
    running with core module 0.0.0 (develop)
    the reference module is  0.0.0 (develop)
    the OpenMP    module is  0.0.0 (develop)
    the CUDA      module is  0.0.0 (develop)
terminate called after throwing an instance of 'gko::CublasError'
  what():  /home/gflegar/Sources/ginkgo/public/repo/cuda/base/cublas_bindings.hpp:198: norm2: CUBLAS_STATUS_NOT_SUPPORTED
zsh: abort (core dumped)  ./inverse_iteration cuda
```